File: 044_Pentagon_Numbers/Solution.go
package main

import (
	"fmt"
	"math"
)

func pentagonNum(num int) int {
	return num * (3*num - 1) / 2
}

func isPentagonNum(numP int) bool {
	// Using Quadratic eqn
	a := 3
	b := -1
	c := -2
	D := int(math.Sqrt(float64(b*b - 4*a*c*numP)))
	x := 2 * a
	rt1 := (-b + D) / x
	rt2 := (-b - D) / x

	if rt1 > 0 && numP == pentagonNum(rt1) {
		return true
	} else if rt2 > 0 && numP == pentagonNum(rt2) {
		return true
	} else {
		return false
	}
}

func main() {
	var n, k int
	fmt.Scan(&n, &k)

	for num := k + 1; num < n; num++ {
		if isPentagonNum(pentagonNum(num) - pentagonNum(num - k)) || isPentagonNum(pentagonNum(num) + pentagonNum(num - k)) {
			fmt.Println(pentagonNum(num))
		}
	}
}

File: 044_Pentagon_Numbers/Solution.py
import math

def pentagon_num(num):
    return num * (3 * num - 1) // 2

def is_pentagon_num(num_p):
    # Using Quadratic eqn
    a = 3
    b = -1
    c = -2
    D = int(math.sqrt(b**2 - 4 * a * c * num_p))
    x = 2 * a
    rt_1 = (-b + D) // x
    rt_2 = (-b - D) // x
    
    if rt_1 > 0 and num_p == pentagon_num(rt_1):
        return True
    elif rt_2 > 0 and num_p == pentagon_num(rt_2):
        return True
    else:
        return False

n, k = map(int, input().strip().split())

for num in range(k +1 , n):
    if is_pentagon_num(pentagon_num(num) - pentagon_num(num - k)) or is_pentagon_num(pentagon_num(num) + pentagon_num(num - k)):
        print(pentagon_num(num))
    
    